internal/dynamic/plugin/input.go

```diff
@@ -0,0 +1,212 @@
+		if err != nil {
+			return nil, err
+		}
+		if spec.Env == nil {
+			spec.Env = make(map[string]string)
+		}
+		socketPath, err := newUnixSocketAddr()
+		if err != nil {
+			return nil, err
+		}
+		var cleanup []func() error
+		defer func() {
+			for _, fn := range cleanup {
+				err := fn()
+				if err != nil {
+					res.Logger().Warnf("failed to clean up creating %s: %v", spec.Name, err)
+				}
+			}
+		}()
+		// No I/O happens in NewClient, so we can do this before we start the subprocess.
+		// This simplifies the cleanup if there is a failure.
+		conn, err := grpc.NewClient(
+			socketPath,
+			grpc.WithTransportCredentials(insecure.NewCredentials()),
+		)
+		if err != nil {
+			return nil, err
+		}
+		cleanup = append(cleanup, conn.Close)
+		spec.Env["REDPANDA_CONNECT_PLUGIN_ADDRESS"] = socketPath
+		proc, err := subprocess.New(spec.Cmd, spec.Env, subprocess.WithLogger(res.Logger()))
+		if err := proc.Start(); err != nil {
+			return nil, fmt.Errorf("unable to start subprocess: %w", err)
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), maxStartupTime)
+		defer cancel()
+		cleanup = append(cleanup, func() error {
+			return proc.Close(ctx)
+		})
+		client := runtimepb.NewBatchInputServiceClient(conn)
+		autoRetryNacks, err := startInputPlugin(ctx, proc, client, cfgValue)
+		if err != nil {
+			return nil, fmt.Errorf("unable to restart plugin: %w", err)
+		}
+		i := &input{
+			cfgValue: cfgValue,
+			proc:     proc,
+			client:   client,
+		}
+		cleanup = nil // Prevent cleanup from running.
+		if autoRetryNacks {
+			return service.AutoRetryNacksBatched(i), nil
+		}
+		return i, nil
+	}
+	return env.RegisterBatchInput(spec.Name, spec.Spec, ctor)
+}
+
+func startInputPlugin(
+	ctx context.Context,
+	proc subprocess.SubProcess,
+	client runtimepb.BatchInputServiceClient,
+	cfgValue any,
+) (autoRetryNacks bool, err error) {
+	if err := proc.Start(); err != nil {
+		return false, fmt.Errorf("unable to restart plugin: %w", err)
+	}
+	value, err := runtimepb.AnyToProto(cfgValue)
+	if err != nil {
+		return false, fmt.Errorf("unable to convert config to proto: %w", err)
+	}
+	autoRetryNacks, err = backoff.RetryWithData(func() (bool, error) {
+		resp, err := client.Init(ctx, &runtimepb.BatchInputInitRequest{
+			Config: value,
+		})
+		if err != nil {
+			if !proc.IsRunning() {
+				return false, backoff.Permanent(fmt.Errorf("plugin exited early: %w", err))
+			}
+			return false, err
+		}
+		if err = runtimepb.ProtoToError(resp.Error); err != nil {
+			return false, backoff.Permanent(err)
+		}
+		return resp.AutoReplayNacks, nil
+	}, backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(maxStartupTime)))
+	if err != nil {
+		return false, fmt.Errorf("unable to initialize plugin: %w", err)
+	}
+	return autoRetryNacks, nil
+}
+
+// Connect implements service.BatchInput.
+func (i *input) Connect(ctx context.Context) (err error) {
+	var resp *runtimepb.BatchInputConnectResponse
+	for range retryCount {
```

Comment:
Would be nice to document why we want this ad-hoc retry here.

Reply:
So this is to wait for plugin to start and accept connections isn't it?

Reply:
Added comments